c10/util/Optional.h

```diff
@@ -238,23 +308,107 @@ struct constexpr_optional_base {
       : init_(true), storage_(il, std::forward<Args>(args)...) {}
 
   ~constexpr_optional_base() = default;
+
+  constexpr_optional_base& operator=(const constexpr_optional_base& rhs) {
+    if (init_ && !rhs.init_) {
+      clear();
+    } else if (!init_ && rhs.init_) {
+      init_ = true;
+      ::new (dataptr()) T(rhs.storage_.value_);
+    } else if (init_ && rhs.init_) {
+      storage_.value_ = rhs.storage_.value_;
+    }
+    return *this;
+  }
+
+  constexpr_optional_base& operator=(constexpr_optional_base&& rhs) noexcept(
+      std::is_nothrow_move_assignable<T>::value &&
+          std::is_nothrow_move_constructible<T>::value) {
+    if (init_ && !rhs.init_) {
+      clear();
+    } else if (!init_ && rhs.init_) {
+      init_ = true;
+      ::new (dataptr()) T(std::move(rhs.storage_.value_));
+    } else if (init_ && rhs.init_) {
+      storage_.value_ = std::move(rhs.storage_.value_);
+    }
+    return *this;
+  }
+
+ private:
+  typename std::remove_const<T>::type* dataptr() {
+    return std::addressof(storage_.value_);
+  }
+
+  constexpr const T* dataptr() const {
+    return detail_::static_addressof(storage_.value_);
+  }
+
+  void clear() noexcept {
+    init_ = false;
+  }
+};
+
+// HACK: Optimization for small scalars. The mainline implementation
+// fails to have trivial copy/move operations in these cases, and we
+// care about them, so just implement that directly.
+template <class T>
+struct small_scalar_optimization_optional_base {
+  bool init_;
+  struct Storage {
+    T value_;
+
+    Storage() = default;
+    explicit Storage(T v) : value_(v) {}
+  } storage_;
+
+  constexpr small_scalar_optimization_optional_base() noexcept
+      : init_(false) {}
+
+  explicit constexpr small_scalar_optimization_optional_base(const T& v)
+      : init_(true), storage_(v) {}
+
+  explicit constexpr small_scalar_optimization_optional_base(T&& v)
+      : init_(true), storage_(constexpr_move(v)) {}
+
+  template <class... Args>
+  explicit constexpr small_scalar_optimization_optional_base(in_place_t, Args&&... args)
+      : init_(true), storage_(constexpr_forward<Args>(args)...) {}
+
+  template <
+      class U,
+      class... Args,
+      TR2_OPTIONAL_REQUIRES(std::is_constructible<T, std::initializer_list<U>>)>
+  constexpr explicit small_scalar_optimization_optional_base(
+      in_place_t,
+      std::initializer_list<U> il,
+      Args&&... args)
+      : init_(true), storage_(il, std::forward<Args>(args)...) {}
+
+  ~small_scalar_optimization_optional_base() = default;
 };
 
 template <class T>
 using OptionalBase = typename std::conditional<
-    std::is_trivially_destructible<T>::value, // if possible
-    constexpr_optional_base<typename std::remove_const<
-        T>::type>, // use base with trivial destructor
-    optional_base<typename std::remove_const<T>::type>>::type;
+    std::is_scalar<T>::value && sizeof(T) <= 4,
```

Comment:
`c10::ScalarType` and `c10::Layout` are enums, so they'll hit the optimization here, but we'd also like to hit the optimization for `c10::Device`, which is a 32bit struct that can for most purposes be treated like a Scalar.

Should we use checks for trivial movability/copyability here instead of `std::is_scalar`? That might hit `c10::Device` too. Also not sure if we need the 32bit sizeof constraint. If a type is trivially movable/copyable, then the optimization could make sense even if the type is larger.

We might need to rewrite the optimization a bit though. What it currently does is it unconditionally calls the copy/move constructor, possibly on uninitialized elements. With scalars, that is likely within the standard. I'm not sure if doing that for trivially copy/movable types is within the standard or UB. Do you happen to know?

Reply:
> it unconditionally calls the copy/move constructor, possibly on uninitialized elements. With scalars, that is likely within the standard. I'm not sure if doing that for trivially copy/movable types is within the standard or UB. Do you happen to know?

We're not using a union, so all we require is that we can copy and move default-constructed objects of type T. If a type can't do that, it's broken. I'm not sure what's possibly uninitialized by the optional implementation; initialization is left up to T's default constructor.

Reply:
> We're not using a union

Ah, and that's a problem because T may not have a default constructor. I'll see what I can do.

Reply:
I feel like, for all the types we care about this optimization applying to, we have default constructors